### PR TITLE
Pass `Context` as `&mut` in rustdoc

### PR DESCRIPTION
**Pull request by Milo123459**
Solves https://github.com/rust-lang/rust/issues/90323

### CODE REVIEW

**Review by Korabi-dev — APPROVED** [review]

File: src/librustdoc/html/render/mod.rs [hunk 1]
```diff
@@ -489,19 +489,19 @@ fn document(
...
-    let mut ids = cx.id_map.borrow_mut();
+    let ids = &mut cx.id_map;
     write!(
         w,
         "<div class=\"docblock\">{}</div>",
         Markdown {
             content: md_text,
             links: &links,
-            ids: &mut ids,
+            ids: ids,
```
Comment (jyn514) [inline src/librustdoc/html/render/mod.rs:504]:
```suggestion
            ids,
```

File: src/librustdoc/html/render/mod.rs [hunk 2]
```diff
@@ -706,7 +700,12 @@ fn short_item_info(
 
 // Render the list of items inside one of the sections "Trait Implementations",
 // "Auto Trait Implementations," "Blanket Trait Implementations" (on struct/enum pages).
-fn render_impls(cx: &Context<'_>, w: &mut Buffer, impls: &[&&Impl], containing_item: &clean::Item) {
+fn render_impls(
+    cx: &mut Context<'_>,
+    w: &mut Buffer,
+    impls: &[&&Impl],
```
Comment (jyn514) [inline src/librustdoc/html/render/mod.rs:706]: Lol this is some very strange code ... It's preexisting though, I'll open an issue.

File: src/librustdoc/html/render/mod.rs [hunk 3]
```diff
@@ -765,13 +764,13 @@ fn assoc_const(
...
     extra: &str,
-    cx: &Context<'_>,
+    cx: &mut Context<'_>,
 ) {
     write!(
         w,
         "{}{}const <a href=\"{}\" class=\"constant\">{}</a>: {}",
         extra,
-        it.visibility.print_with_space(it.def_id, cx),
+        it.visibility.print_with_space(it.def_id, &cx.clone()),
```
Comment (jyn514) [inline src/librustdoc/html/render/mod.rs:773]: Why is this clone necessary? I think you can just do `&*cx`.

Comment (Milo123459) [reply src/librustdoc/html/render/mod.rs:773]: Not sure, let me test

File: src/librustdoc/html/render/mod.rs (hunk 2 above)
Comment (Milo123459) [reply src/librustdoc/html/render/mod.rs:706]: I mean, I only changed how `cx` was defined.

File: src/librustdoc/html/render/mod.rs (hunk 3 above)
Comment (Milo123459) [reply src/librustdoc/html/render/mod.rs:773]: Ah right, it differs the mutability and causes borrowing issues.

**Comment by jyn514** [comment]
@Milo123459 I suspect the test failure is because of all the new clones you added and that rustdoc depended on the state being shared. Do you think it would be feasible to remove the clones? I can help you out if you run into trouble :)

**Comment by Milo123459** [comment]
Sure! Thanks

File: src/librustdoc/html/render/mod.rs (hunk 3 above)
Comment (fee1-dead) [reply src/librustdoc/html/render/mod.rs:773]: What about moving this up and before the `write` so you can use `&mut cx` after?

File: src/librustdoc/html/render/context.rs [hunk 4]
```diff
@@ -43,6 +43,8 @@ use crate::try_err;
 /// It is intended that this context is a lightweight object which can be fairly
 /// easily cloned because it is cloned per work-job (about once per item in the
 /// rustdoc tree).
+
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/context.rs:46]: Please remove this empty line (it looks super weird to have a doc comment not followed directly by the item).

File: src/librustdoc/html/render/context.rs [hunk 5]
```diff
@@ -43,6 +43,7 @@ use crate::try_err;
 /// It is intended that this context is a lightweight object which can be fairly
 /// easily cloned because it is cloned per work-job (about once per item in the
 /// rustdoc tree).
+#[derive(Clone)]
```
Comment (camelid) [inline src/librustdoc/html/render/context.rs:46]: This must be removed before this PR is merged; `Context` intentionally does not implement Clone (cf. `make_child_renderer()`). We should probably add something to its docs about this.

**Review by camelid — CHANGES_REQUESTED** [review]

File: src/librustdoc/html/render/context.rs (hunk 5 above)
Comment (Milo123459) [reply src/librustdoc/html/render/context.rs:46]: Alright, shall work on it.

Comment (camelid) [reply src/librustdoc/html/render/context.rs:46]: Please feel free to ask for help here or on Zulip :)

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/context.rs:46]: An easy to check that a type doesn't implement a trait is to have a doc code like this:

``````rust
/// ```compile_fail
/// let c = Context { ... };
/// c.clone();
/// ```
``````

:)

Comment (Milo123459) [reply src/librustdoc/html/render/context.rs:46]: Hey, are you suggesting that has been implemented or should be implemented?

Anyway - I'm almost finished removing the clones!

Comment (Milo123459) [reply src/librustdoc/html/render/context.rs:46]: Yea I take that back, this is proving to be a challenge haha

**Pull request closed without merging**